pybaseball/datasources/fangraphs.py

```diff
@@ -1,44 +1,135 @@
+from abc import ABC
+from enum import Enum, unique
+from typing import Callable, Iterable, List, Optional, Type, Union
 
 import lxml
 import pandas as pd
 import requests
 
 from pybaseball.datahelpers import postprocessing
+from pybaseball.datahelpers.column_mapper import BattingStatsColumnMapper, GenericColumnMapper, ColumnListMapperFunction
+from pybaseball.datasources.html_table_processor import HTMLTableProcessor
+from pybaseball.enums.fangraphs import (FanGraphsStat, FanGraphsBattingStat, FanGraphsFieldingStat, FanGraphsLeague, FanGraphsMonth,
+                                        FanGraphsPitchingStat, FanGraphsPositions, FanGraphsStatTypes)
 
-ROOT_URL = 'https://www.fangraphs.com'
+_FG_LEADERS_URL = "/leaders.aspx"
 
+MIN_AGE = 0
+MAX_AGE = 100
 
-def get_fangraphs_tabular_data_from_html(html: Union[str, bytes], column_name_mapper: Callable = None, known_percentages: List[str] = []) -> pd.DataFrame:
-    xpath: str = '//table[@class="rgMasterTable"]'
-    html_dom = lxml.etree.HTML(html)
+class FanGraphsDataTable(ABC):
+    ROOT_URL: str = "https://www.fangraphs.com"
+    TABLE_CLASS: str = "rgMasterTable"
+    HEADINGS_XPATH: str = "({TABLE_XPATH}/thead//th[contains(@class, 'rgHeader')])[position()>1]/descendant-or-self::*/text()"
+    DATA_ROWS_XPATH: str = "({TABLE_XPATH}/tbody//tr)"
+    DATA_CELLS_XPATH: str = "td[position()>1]/descendant-or-self::*/text()"
+    QUERY_ENDPOINT: str = _FG_LEADERS_URL
+    STATS: FanGraphsStatTypes = FanGraphsStatTypes.NONE
+    DEFAULT_TYPES: FanGraphsStat = None
+    KNOWN_PERCENTAGES: List[str] = []
+    TEAM_DATA: bool = False
+    COLUMN_NAME_MAPPER: ColumnListMapperFunction = GenericColumnMapper().map_list
 
-    headings_xpath = f"({xpath}/thead//th[contains(@class, 'rgHeader')])[position()>1]/descendant-or-self::*/text()"
-    headings = html_dom.xpath(headings_xpath)
+    def __init__(self):
+        self.html_accessor = HTMLTableProcessor(
+            root_url=self.ROOT_URL,
+            headings_xpath=self.HEADINGS_XPATH,
+            data_cell_xpath=self.DATA_CELLS_XPATH,
+            data_rows_xpath=self.DATA_ROWS_XPATH,
+            table_class=self.TABLE_CLASS,
+        )
 
-    if column_name_mapper:
-        headings = [column_name_mapper(h) for h in headings]
+    def _postprocess(self, data: pd.DataFrame) -> pd.DataFrame:
+        return data
 
-    data_rows_xpath = f"({xpath}/tbody//tr)"
-    data_rows_dom = html_dom.xpath(data_rows_xpath)
-    data_rows = [
-        [
-            postprocessing.try_parse(y, headings[index], known_percentages=known_percentages)
-            for index, y in enumerate(x.xpath('td[position()>1]/descendant-or-self::*/text()'))
-        ]
-        for x in data_rows_dom
-    ]
+    def _validate(self, data: pd.DataFrame) -> pd.DataFrame:
+        return data
 
-    fg_data = pd.DataFrame(data_rows, columns=headings)
+    def fetch(self, start_season: int, end_season: Optional[int] = None, types: FanGraphsStat = None,
```

Comment:
I know I'm the one who removed it in my example PR, but could we put the doc string back in here?

Reply:
CURSE YOU @bdilday.

Yeah I'll add it back in. I had intended to anyway.